Logging of whitelist/blacklist failures and slight refactor

diginc opened the pull request:
Helps with debugging of issues like https://github.com/pi-hole/pi-hole/issues/546 and https://github.com/pi-hole/AdminLTE/issues/100 - I saw another new report of issues in the 546 ticket and already had most of this work done.

- Adds logging of failure cases for CORS
- Makes the authorized hosts domains for CORS an array to simplify adding more in the future (like virtualhost aliases)

When I cut the 1.7.0 docker release I debugged why CORS was failing in my docker setup (ended up being the same reason as #100, custom hostname) and added SERVER_NAME var for a fix.  But then I realized SERVER_NAME php var was highly insecure in the Lighttpd's and Apache's implementation, so that is removed from this but I left in my logging of failures.  See https://github.com/pi-hole/AdminLTE/issues/100#issuecomment-226678820 for more info on SERVER_NAME, there are also a bunch of good stackoverflow and blog articles about $SERVER_NAME security..

I've only tested this in my docker of 1.7.0 so it probably needs some more testing.

@pi-hole/dashboard

Mcat12 commented on php/sub.php:
```diff
@@ -1,26 +1,49 @@
 
 switch($_POST['list']) {
     case "white":
-        exec("sudo pihole -w -q -d ${_POST['domain']}");
```
The `-d` flags need to be added back in.

Mcat12 commented:
The CORS check is sort of working, but not... I was able to send a request from another server to sub.php and got to the "Wrong token" result (I set token to "none"), which meant I had passed CORS. However, running it in a real web page through JS had Chrome yelling at me as if the check worked. Basically, the line `die_and_log("Failed CORS");` is never called, but it still works because the CORS header it puts out doesn't match the origin of the requester. We should actually stop checking the Host header, since that is actually the address of the resource being requested, not related to the origin (my fault). Also, `$_SERVER['HTTP_ORIGIN']` could include `http://`, so the CORS check would fail even if the origin was correct.

It looks like both this PR and the current release has that bug (checking Host header)... Good thing we're using CSRF also!

diginc commented:
I refactored some more last night to integrate #112 (the fix you mentioned)

I also added some improvements to logging and hopefully the ENV for $VIRTUAL_HOST will address #100 (custom internal hostname/FQDN like pihole.domain.local not working).  I did not get to successfully pass $VIRTUAL_HOST into lighttpd's php config environment variables while testing though.  I believe I was doing it right but when I printed $APPROVED_HOSTNAMES in the log it showed it wasn't there yet so there is still some work to do on this to prove out that part is working.

diginc commented:
I've confirmed #100 is fixed now and managed to get PHP envs vars working for debian+fastcgi and alpine+php-fpm in my docker image.

Take a look at https://github.com/diginc/docker-pi-hole/pull/17/files to see some of the php config mods needed to pass down Env Vars for debian and alpine.  My docker image now has a alpine_dev and debian_dev image this fork's changesets baked into the image if you perfer to test that way.  https://hub.docker.com/r/diginc/pi-hole/tags/

diginc commented:
For what it's worth this is in my docker container's latest version and has been for almost a week.

Mcat12 commented on php/auth.php:
```diff
@@ -0,0 +1,49 @@
+<?php
+$ERRORLOG = getenv('PHP_ERROR_LOG');
```
Does this variable need to be set? PHP is yelling that the file name on line 5 is empty, which I assume would mean that this is empty.

Mcat12 replied:
Would `ini_get('error_log');` work the same? I would rather not have to make the web interface wait for an update from Pi-hole Core.

Mcat12 replied:
That `ini_get` doesn't work either. I propose setting `$ERRORLOG` equal to `"/var/log/nginx/pihole_error.log"` or something that would not rely on Pi-hole Core.

diginc replied:
Good call, I forgot to add a default at this layer so core pi-hole doesn't need to worry about php ENVs.  My docker does that for me :(  I'll add some checking logic to set a default if not found as an ENV/SERVER var

diginc replied:
isn't lighttpd still the default web server for pi-hole core or did they switch to NGINX and I missed that?  I was thinking /var/log/lighttpd/error.log (mixed in with the lighttpd log, I formatted the date string the same as lighttpd I believe)

Mcat12 replied:
Oh ya. Getting mixed up with the python rewrite (since on my dev machine I removed lighttpd, php can only write to /var/log/nginx/).

Mcat12 commented on php/auth.php:
```diff
@@ -0,0 +1,52 @@
+$ERRORLOG = getenv('PHP_ERROR_LOG');
+if (empty($ERRORLOG)) {
+    $ERRORLOG = '/var/log/lighttpd/error.log'
+}
```
Syntax error on line 4. No `;`

Mcat12 commented:
Approved

PromoFaux commented:
Approved!

brantje commented:
Code looks good, approved.
